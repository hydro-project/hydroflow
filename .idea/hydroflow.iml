<?xml version="1.0" encoding="UTF-8"?>
<module type="JAVA_MODULE" version="4">
  <component name="NewModuleRootManager" inherit-compiler-output="true">
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/benches/benches" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/hydro_deploy/core/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydro_deploy/hydro_cli/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydro_deploy/hydro_cli_examples/examples" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydro_deploy/hydroflow_deploy_integration/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydro_deploy/hydroflow_plus_deploy/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow/examples" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_datalog/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_datalog_core/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_lang/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_macro/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_plus/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_plus_test/examples" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_plus_test/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/lattices/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/lattices/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/lattices_macro/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/multiplatform_test/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/multiplatform_test/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/pusherator/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/relalg/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/stageleft/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/stageleft_macro/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/stageleft_test/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/stageleft_tool/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/topolotree/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/variadics/examples" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/variadics/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/variadics/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/website_playground/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/website_playground/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/datastores/gossip_kv/examples" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/datastores/gossip_kv/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/hydroflow_plus_test_local/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/target" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>