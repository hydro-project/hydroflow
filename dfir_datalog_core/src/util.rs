use std::ops::RangeFrom;

use quote::ToTokens;
use syn::parse_quote;

pub type Counter = RangeFrom<isize>;

pub fn repeat_tuple<I: ToTokens, T: syn::parse::Parse>(rep: fn() -> I, n: usize) -> T {
    let values = (1..=n).map(|_| rep()).collect::<Vec<I>>();

    parse_quote!((#(#values, )*))
}
