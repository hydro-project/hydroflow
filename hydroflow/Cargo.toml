[package]
name = "hydroflow"
publish = true
version = "0.6.2"
edition = "2021"
license = "Apache-2.0"
documentation = "https://docs.rs/hydroflow/"
description = "Hydro's low-level dataflow runtime and IR"

[features]
default = [ "macros" , "nightly", "debugging" ]

nightly = [ "hydroflow_macro", "hydroflow_macro/diagnostics" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]
hydroflow_macro = [ "dep:hydroflow_macro" ]
hydroflow_datalog = [ "dep:hydroflow_datalog" ]
cli_integration = [ "dep:hydroflow_cli_integration" ]
python = [ "dep:pyo3" ]
debugging = [ "hydroflow_lang/debugging" ]

[[example]]
name = "kvs_bench"
required-features = [ "nightly" ]

[[example]]
name = "python_udf"
required-features = [ "python" ]

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = "0.3"
hydroflow_cli_integration = { optional = true, path = "../hydro_deploy/hydroflow_cli_integration", version = "^0.5.2" }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog", version = "^0.6.0" }
hydroflow_lang = { path = "../hydroflow_lang", version = "^0.6.2" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro", version = "^0.6.0" }
itertools = "0.10"
lattices = { path = "../lattices", version = "^0.5.4", features = [ "serde" ] }
pusherator = { path = "../pusherator", version = "^0.0.5" }
pyo3 = { optional = true, version = "0.20" }
ref-cast = "1.0"
regex = "1.8.4"
rustc-hash = "1.1.0"
sealed = "0.5"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
smallvec = "1.10.0"
tokio-stream = { version = "0.1.10", features = [ "io-util", "sync" ] }
tracing = "0.1"
variadics = { path = "../variadics", version = "^0.0.4" }
web-time = "1.1.0"

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
tokio = { version = "1.16", features = [ "full" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }

[target.'cfg(target_arch = "wasm32")'.dependencies]
tokio = { version = "1.16", features = [ "rt" , "sync", "macros", "io-util", "time" ] }
tokio-util = { version = "0.7.4", features = [ "codec" ] }
# We depend on getrandom transitively through rand. To compile getrandom to
# WASM, we need to enable its "js" feature. However, rand does not expose a
# passthrough to enable "js" on getrandom. As a workaround, we enable the
# getrandom js feature here; when the feature resolver computes the set of
# features for getrandom (unification), it will include "js".
getrandom = { version = "0.2.6", features = [ "js" ] }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde", "clock" ], default-features = false }
clap = { version = "4.1.8", features = [ "derive" ] }
colored = "2.0"
core_affinity = "0.5.10"
ctor = "0.2"
futures = { version = "0.3" }
hdrhistogram = "7"
insta = "1.7.1"
multiplatform_test = { path = "../multiplatform_test", version = "0.0.0" }
wasm-bindgen-test = "0.3.34"
rand = {version = "0.8.4", features = ["small_rng"]}
rand_distr = "0.4.3"
regex = "1"
static_assertions = "1.1.0"
textnonce = "1.0.0"
time = "0.3"
tracing-subscriber = { version = "0.3", features = [ "env-filter" ] }
trybuild = "1.0.80"
zipf = "7.0.0"

[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
# Rayon (rust data-parallelism library) does not compile on WASM.
criterion = { version = "0.5", features = [ "async_tokio", "html_reports" ] }

