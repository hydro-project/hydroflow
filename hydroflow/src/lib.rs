#![warn(missing_docs)]

//! Hydroflow is a low-level dataflow-based runtime system for the [Hydro Project](https://hydro.run/).
//!
//! The primary item in this crate is the [`Hydroflow`](crate::scheduled::graph::Hydroflow) struct,
//! representing a Hydroflow dataflow graph. Although this graph can be manually constructed, the
//! easiest way to instantiate a `Hydroflow` instance is with the [`hydroflow_syntax!`] macro using
//! Hydroflow's custom "surface syntax."
//!
//! ```rust
//! let mut hf = hydroflow::hydroflow_syntax! {
//!     source_iter(["hello", "world"]) -> for_each(|s| println!("{}", s));
//! };
//! hf.run_available();
//! ```
//!
//! For more examples, check out the [`examples` folder on Github](https://github.com/hydro-project/hydroflow/tree/main/hydroflow/examples).

pub mod compiled;
pub mod scheduled;
pub mod util;

#[cfg(feature = "python")]
pub use pyo3;
pub use variadics::{self, var_args, var_expr, var_type};
pub use {
    bincode, bytes, futures, hydroflow_lang as lang, itertools, lattices, pusherator, rustc_hash,
    serde, serde_json, tokio, tokio_stream, tokio_util, tracing, web_time,
};

/// `#[macro_use]` automagically brings the declarative macro export to the crate-level.
mod declarative_macro;
#[cfg(feature = "hydroflow_datalog")]
pub use hydroflow_datalog::*;
#[cfg(feature = "hydroflow_macro")]
pub use hydroflow_macro::{
    hydroflow_main as main, hydroflow_parser, hydroflow_syntax, hydroflow_syntax_noemit,
    hydroflow_test as test, monotonic_fn, morphism, DemuxEnum,
};

// TODO(mingwei): Use the [nightly "never" type `!`](https://doc.rust-lang.org/std/primitive.never.html)
/// Stand-in for the [nightly "never" type `!`](https://doc.rust-lang.org/std/primitive.never.html)
pub type Never = std::convert::Infallible;

#[cfg(doctest)]
mod booktest {
    mod surface_ops {
        hydroflow_macro::surface_booktest_operators!();
    }
}
