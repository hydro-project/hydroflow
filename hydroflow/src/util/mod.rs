#![warn(missing_docs)]
//! Helper utilities for the Hydroflow surface syntax.

pub mod clear;
#[cfg(feature = "hydroflow_macro")]
pub mod demux_enum;
pub mod monotonic_map;
pub mod multiset;
pub mod sparse_vec;
pub mod unsync;

pub mod simulation;

mod monotonic;
pub use monotonic::*;

mod udp;
#[cfg(not(target_arch = "wasm32"))]
pub use udp::*;

mod tcp;
#[cfg(not(target_arch = "wasm32"))]
pub use tcp::*;

#[cfg(unix)]
mod socket;
#[cfg(unix)]
pub use socket::*;

#[cfg(feature = "deploy_integration")]
pub mod deploy;

use std::io::Read;
use std::net::SocketAddr;
use std::num::NonZeroUsize;
use std::process::{Child, ChildStdin, ChildStdout, Stdio};
use std::task::{Context, Poll};

use futures::Stream;
use serde::de::DeserializeOwned;
use serde::ser::Serialize;

/// Persit or delete tuples
pub enum Persistence<T> {
    /// Persist T values
    Persist(T),
    /// Delete all values that exactly match
    Delete(T),
}

/// Persit or delete key-value pairs
pub enum PersistenceKeyed<K, V> {
    /// Persist key-value pairs
    Persist(K, V),
    /// Delete all tuples that have the key K
    Delete(K),
}

/// Returns a channel as a (1) unbounded sender and (2) unbounded receiver `Stream` for use in Hydroflow.
pub fn unbounded_channel<T>() -> (
    tokio::sync::mpsc::UnboundedSender<T>,
    tokio_stream::wrappers::UnboundedReceiverStream<T>,
) {
    let (send, recv) = tokio::sync::mpsc::unbounded_channel();
    let recv = tokio_stream::wrappers::UnboundedReceiverStream::new(recv);
    (send, recv)
}

/// Returns an unsync channel as a (1) sender and (2) receiver `Stream` for use in Hydroflow.
pub fn unsync_channel<T>(
    capacity: Option<NonZeroUsize>,
) -> (unsync::mpsc::Sender<T>, unsync::mpsc::Receiver<T>) {
    unsync::mpsc::channel(capacity)
}

/// Returns an [`Iterator`] of any immediately available items from the [`Stream`].
pub fn ready_iter<S>(stream: S) -> impl Iterator<Item = S::Item>
where
    S: Stream,
{
    let mut stream = Box::pin(stream);
    std::iter::from_fn(move || {
        match stream
            .as_mut()
            .poll_next(&mut Context::from_waker(futures::task::noop_waker_ref()))
        {
            Poll::Ready(opt) => opt,
            Poll::Pending => None,
        }
    })
}

/// Collects the immediately available items from the `Stream` into a `FromIterator` collection.
///
/// This consumes the stream, use [`futures::StreamExt::by_ref()`] (or just `&mut ...`) if you want
/// to retain ownership of your stream.
pub fn collect_ready<C, S>(stream: S) -> C
where
    C: FromIterator<S::Item>,
    S: Stream,
{
    assert!(tokio::runtime::Handle::try_current().is_err(), "Calling `collect_ready` from an async runtime may cause incorrect results, use `collect_ready_async` instead.");
    ready_iter(stream).collect()
}

/// Collects the immediately available items from the `Stream` into a collection (`Default` + `Extend`).
///
/// This consumes the stream, use [`futures::StreamExt::by_ref()`] (or just `&mut ...`) if you want
/// to retain ownership of your stream.
pub async fn collect_ready_async<C, S>(stream: S) -> C
where
    C: Default + Extend<S::Item>,
    S: Stream,
{
    use std::sync::atomic::Ordering;

    // Yield to let any background async tasks send to the stream.
    tokio::task::yield_now().await;

    let got_any_items = std::sync::atomic::AtomicBool::new(true);
    let mut unfused_iter =
        ready_iter(stream).inspect(|_| got_any_items.store(true, Ordering::Relaxed));
    let mut out = C::default();
    while got_any_items.swap(false, Ordering::Relaxed) {
        out.extend(unfused_iter.by_ref());
        // Tokio unbounded channel returns items in lenght-128 chunks, so we have to be careful
        // that everything gets returned. That is why we yield here and loop.
        tokio::task::yield_now().await;
    }
    out
}

/// Serialize a message to bytes using bincode.
pub fn serialize_to_bytes<T>(msg: T) -> bytes::Bytes
where
    T: Serialize,
{
    bytes::Bytes::from(bincode::serialize(&msg).unwrap())
}

/// Serialize a message from bytes using bincode.
pub fn deserialize_from_bytes<T>(msg: impl AsRef<[u8]>) -> bincode::Result<T>
where
    T: DeserializeOwned,
{
    bincode::deserialize(msg.as_ref())
}

/// Resolve the `ipv4` [`SocketAddr`] from an IP or hostname string.
pub fn ipv4_resolve(addr: &str) -> Result<SocketAddr, std::io::Error> {
    use std::net::ToSocketAddrs;
    let mut addrs = addr.to_socket_addrs()?;
    let result = addrs.find(|addr| addr.is_ipv4());
    match result {
        Some(addr) => Ok(addr),
        None => Err(std::io::Error::new(
            std::io::ErrorKind::Other,
            "Unable to resolve IPv4 address",
        )),
    }
}

/// Returns a length-delimited bytes `Sink`, `Stream`, and `SocketAddr` bound to the given address.
/// The input `addr` may have a port of `0`, the returned `SocketAddr` will have the chosen port.
#[cfg(not(target_arch = "wasm32"))]
pub async fn bind_udp_bytes(addr: SocketAddr) -> (UdpSink, UdpStream, SocketAddr) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_bytes(socket)
}

/// Returns a newline-delimited bytes `Sink`, `Stream`, and `SocketAddr` bound to the given address.
/// The input `addr` may have a port of `0`, the returned `SocketAddr` will have the chosen port.
#[cfg(not(target_arch = "wasm32"))]
pub async fn bind_udp_lines(addr: SocketAddr) -> (UdpLinesSink, UdpLinesStream, SocketAddr) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_lines(socket)
}

/// Returns a newline-delimited bytes `Sender`, `Receiver`, and `SocketAddr` bound to the given address.
///
/// The input `addr` may have a port of `0`, the returned `SocketAddr` will be the address of the newly bound endpoint.
/// The inbound connections can be used in full duplex mode. When a `(T, SocketAddr)` pair is fed to the `Sender`
/// returned by this function, the `SocketAddr` will be looked up against the currently existing connections.
/// If a match is found then the data will be sent on that connection. If no match is found then the data is silently dropped.
#[cfg(not(target_arch = "wasm32"))]
pub async fn bind_tcp_bytes(
    addr: SocketAddr,
) -> (
    unsync::mpsc::Sender<(bytes::Bytes, SocketAddr)>,
    unsync::mpsc::Receiver<Result<(bytes::BytesMut, SocketAddr), std::io::Error>>,
    SocketAddr,
) {
    bind_tcp(addr, tokio_util::codec::LengthDelimitedCodec::new())
        .await
        .unwrap()
}

/// This is the same thing as `bind_tcp_bytes` except instead of using a length-delimited encoding scheme it uses new lines to separate frames.
#[cfg(not(target_arch = "wasm32"))]
pub async fn bind_tcp_lines(
    addr: SocketAddr,
) -> (
    unsync::mpsc::Sender<(String, SocketAddr)>,
    unsync::mpsc::Receiver<Result<(String, SocketAddr), tokio_util::codec::LinesCodecError>>,
    SocketAddr,
) {
    bind_tcp(addr, tokio_util::codec::LinesCodec::new())
        .await
        .unwrap()
}

/// The inverse of [`bind_tcp_bytes`].
///
/// `(Bytes, SocketAddr)` pairs fed to the returned `Sender` will initiate new tcp connections to the specified `SocketAddr`.
/// These connections will be cached and reused, so that there will only be one connection per destination endpoint. When the endpoint sends data back it will be available via the returned `Receiver`
#[cfg(not(target_arch = "wasm32"))]
pub fn connect_tcp_bytes() -> (
    TcpFramedSink<bytes::Bytes>,
    TcpFramedStream<tokio_util::codec::LengthDelimitedCodec>,
) {
    connect_tcp(tokio_util::codec::LengthDelimitedCodec::new())
}

/// This is the same thing as `connect_tcp_bytes` except instead of using a length-delimited encoding scheme it uses new lines to separate frames.
#[cfg(not(target_arch = "wasm32"))]
pub fn connect_tcp_lines() -> (
    TcpFramedSink<String>,
    TcpFramedStream<tokio_util::codec::LinesCodec>,
) {
    connect_tcp(tokio_util::codec::LinesCodec::new())
}

/// Sort a slice using a key fn which returns references.
///
/// From addendum in
/// <https://stackoverflow.com/questions/56105305/how-to-sort-a-vec-of-structs-by-a-string-field>
pub fn sort_unstable_by_key_hrtb<T, F, K>(slice: &mut [T], f: F)
where
    F: for<'a> Fn(&'a T) -> &'a K,
    K: Ord,
{
    slice.sort_unstable_by(|a, b| f(a).cmp(f(b)))
}

/// Waits for a specific process output before returning.
///
/// When a child process is spawned often you want to wait until the child process is ready before
/// moving on. One way to do that synchronization is by waiting for the child process to output
/// something and match regex against that output. For example, you could wait until the child
/// process outputs "Client live!" which would indicate that it is ready to receive input now on
/// stdin.
pub fn wait_for_process_output(
    output_so_far: &mut String,
    output: &mut ChildStdout,
    wait_for: &str,
) {
    let re = regex::Regex::new(wait_for).unwrap();

    while !re.is_match(output_so_far) {
        println!("waiting: {}", output_so_far);
        let mut buffer = [0u8; 1024];
        let bytes_read = output.read(&mut buffer).unwrap();

        if bytes_read == 0 {
            panic!();
        }

        output_so_far.push_str(&String::from_utf8_lossy(&buffer[0..bytes_read]));

        println!("XXX {}", output_so_far);
    }
}

/// Terminates the inner [`Child`] process when dropped.
///
/// When a `Child` is dropped normally nothing happens but in unit tests you usually want to
/// terminate the child and wait for it to terminate. `DroppableChild` does that for us.
pub struct DroppableChild(Child);

impl Drop for DroppableChild {
    fn drop(&mut self) {
        #[cfg(target_family = "windows")]
        let _ = self.0.kill(); // Windows throws `PermissionDenied` if the process has already exited.
        #[cfg(not(target_family = "windows"))]
        self.0.kill().unwrap();

        self.0.wait().unwrap();
    }
}

/// Run a rust example as a test.
///
/// Rust examples are meant to be run by people and have a natural interface for that. This makes
/// unit testing them cumbersome. This function wraps calling cargo run and piping the stdin/stdout
/// of the example to easy to handle returned objects. The function also returns a `DroppableChild`
/// which will ensure that the child processes will be cleaned up appropriately.
pub fn run_cargo_example(test_name: &str, args: &str) -> (DroppableChild, ChildStdin, ChildStdout) {
    let mut server = if args.is_empty() {
        std::process::Command::new("cargo")
            .args(["run", "-p", "hydroflow", "--example"])
            .arg(test_name)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()
            .unwrap()
    } else {
        std::process::Command::new("cargo")
            .args(["run", "-p", "hydroflow", "--example"])
            .arg(test_name)
            .arg("--")
            .args(args.split(' '))
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()
            .unwrap()
    };

    let stdin = server.stdin.take().unwrap();
    let stdout = server.stdout.take().unwrap();

    (DroppableChild(server), stdin, stdout)
}

/// Converts an iterator into a stream that emits `n` items at a time, yielding between each batch.
///
/// This is useful for breaking up a large iterator across several ticks: `source_iter(...)` always
/// releases all items in the first tick. However using `iter_batches_stream` with `source_stream(...)`
/// will cause `n` items to be released each tick. (Although more than that may be emitted if there
/// are loops in the stratum).
pub fn iter_batches_stream<I>(
    iter: I,
    n: usize,
) -> futures::stream::PollFn<impl FnMut(&mut Context<'_>) -> Poll<Option<I::Item>>>
where
    I: IntoIterator + Unpin,
{
    let mut count = 0;
    let mut iter = iter.into_iter();
    futures::stream::poll_fn(move |ctx| {
        count += 1;
        if n < count {
            count = 0;
            ctx.waker().wake_by_ref();
            Poll::Pending
        } else {
            Poll::Ready(iter.next())
        }
    })
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    pub fn test_collect_ready() {
        let (send, mut recv) = unbounded_channel::<usize>();
        for x in 0..1000 {
            send.send(x).unwrap();
        }
        assert_eq!(1000, collect_ready::<Vec<_>, _>(&mut recv).len());
    }

    #[crate::test]
    pub async fn test_collect_ready_async() {
        // Tokio unbounded channel returns items in 128 item long chunks, so we have to be careful that everything gets returned.
        let (send, mut recv) = unbounded_channel::<usize>();
        for x in 0..1000 {
            send.send(x).unwrap();
        }
        assert_eq!(
            1000,
            collect_ready_async::<Vec<_>, _>(&mut recv).await.len()
        );
    }
}
