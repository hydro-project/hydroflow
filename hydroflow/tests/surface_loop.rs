use hydroflow::util::iter_batches_stream;
use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test]
pub fn test_flo_syntax() {
    let mut df = hydroflow_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>()
                -> map(|item| (context.current_tick().0, item))
                -> assert_eq([
                    (0, ("alice", 0)),
                    (0, ("alice", 1)),
                    (0, ("alice", 2)),
                    (0, ("bob", 0)),
                    (0, ("bob", 1)),
                    (0, ("bob", 2)),
                    (1, ("alice", 3)),
                    (1, ("alice", 4)),
                    (1, ("alice", 5)),
                    (1, ("bob", 3)),
                    (1, ("bob", 4)),
                    (1, ("bob", 5)),
                    (2, ("alice", 6)),
                    (2, ("alice", 7)),
                    (2, ("alice", 8)),
                    (2, ("bob", 6)),
                    (2, ("bob", 7)),
                    (2, ("bob", 8)),
                    (3, ("alice", 9)),
                    (3, ("alice", 10)),
                    (3, ("alice", 11)),
                    (3, ("bob", 9)),
                    (3, ("bob", 10)),
                    (3, ("bob", 11)),
                ]);
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_nested() {
    let mut df = hydroflow_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>();
            loop {
                cp
                    -> all_once()
                    -> map(|vec| (context.current_tick().0, vec))
                    -> assert_eq([
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                    ]);
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_repeat_n() {
    let mut df = hydroflow_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>();
            loop {
                cp
                    -> repeat_n(3)
                    -> map(|vec| (context.current_tick().0, vec))
                    -> inspect(|x| println!("{:?}", x))
                    -> assert_eq([
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                    ]);
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_repeat_n_nested() {
    let mut df = hydroflow_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch() -> flatten();
            loop {
                usrs3 = usrs2 -> repeat_n(3) -> flatten();
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("{:?}", x))
                        -> assert_eq([
                            vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                            vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                            vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                        ]);
                }
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}
