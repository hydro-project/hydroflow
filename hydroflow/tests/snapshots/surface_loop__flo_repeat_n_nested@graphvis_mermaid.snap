---
source: hydroflow/tests/surface_loop.rs
expression: "df.meta_graph().unwrap().to_mermaid(& Default :: default())"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1[\"(1v1) <code>source_iter([&quot;alice&quot;, &quot;bob&quot;])</code>"/]:::pullClass
2v1[\"(2v1) <code>batch()</code>"/]:::pullClass
3v1[\"(3v1) <code>flatten()</code>"/]:::pullClass
4v1[\"(4v1) <code>repeat_n(3)</code>"/]:::pullClass
5v1[\"(5v1) <code>flatten()</code>"/]:::pullClass
6v1[\"(6v1) <code>repeat_n(3)</code>"/]:::pullClass
7v1[\"(7v1) <code>inspect(|x| println!(&quot;{:?}&quot;, x))</code>"/]:::pullClass
8v1[/"<div style=text-align:center>(8v1)</div> <code>assert_eq([<br>    vec![&quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;, &quot;alice&quot;, &quot;bob&quot;],<br>])</code>"\]:::pushClass
9v1["(9v1) <code>handoff</code>"]:::otherClass
10v1["(10v1) <code>handoff</code>"]:::otherClass
11v1["(11v1) <code>handoff</code>"]:::otherClass
12v1["(12v1) <code>handoff</code>"]:::otherClass
13v1["(13v1) <code>handoff</code>"]:::otherClass
2v1-->9v1
1v1-->10v1
4v1-->11v1
3v1-->12v1
7v1-->8v1
6v1-->7v1
5v1-->13v1
9v1-->3v1
10v1-->2v1
11v1-->5v1
12v1--x4v1; linkStyle 10 stroke:red
13v1--x6v1; linkStyle 11 stroke:red
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1
    subgraph sg_1v1_var_usrs1 ["var <tt>usrs1</tt>"]
        1v1
    end
end
subgraph sg_2v1 ["sg_2v1 stratum 0"]
    2v1
    subgraph sg_2v1_var_usrs2 ["var <tt>usrs2</tt>"]
        2v1
    end
end
subgraph sg_3v1 ["sg_3v1 stratum 0"]
    3v1
    subgraph sg_3v1_var_usrs2 ["var <tt>usrs2</tt>"]
        3v1
    end
end
subgraph sg_4v1 ["sg_4v1 stratum 1"]
    4v1
    subgraph sg_4v1_var_usrs3 ["var <tt>usrs3</tt>"]
        4v1
    end
end
subgraph sg_5v1 ["sg_5v1 stratum 1"]
    5v1
    subgraph sg_5v1_var_usrs3 ["var <tt>usrs3</tt>"]
        5v1
    end
end
subgraph sg_6v1 ["sg_6v1 stratum 2"]
    6v1
    7v1
    8v1
end
